python/ray/workflow/workflow_executor.py

```diff
@@ -0,0 +1,366 @@
+
+        # set errors for pending workflow outputs
+        for task_id, futures in self._task_done_callbacks.items():
+            err = ValueError(
+                f"The workflow haven't yet produced output of task '{task_id}' "
+                f"after workflow execution completes."
+            )
+            for fut in futures:
+                if not fut.done():
+                    fut.set_exception(err)
+
+    def cancel(self) -> None:
+        """Cancel the running workflow."""
+        for fut, workflow_ref in self._state.running_frontier.items():
+            fut.cancel()
+            try:
+                ray.cancel(workflow_ref.ref, force=True)
+            except Exception:
+                pass
+
+    def _poll_queued_tasks(self) -> List[StepID]:
+        tasks = []
+        while True:
+            task_id = self._state.pop_frontier_to_run()
+            if task_id is None:
+                break
+            tasks.append(task_id)
+        return tasks
+
+    def _submit_ray_task(self, task_id: StepID, job_id: str) -> None:
+        """Submit a workflow task as a Ray task."""
+        state = self._state
+        baked_inputs = _BakedWorkflowInputs(
+            args=state.task_input_args[task_id],
+            workflow_refs=[
+                state.get_input(d) for d in state.upstream_dependencies[task_id]
+            ],
+        )
+        task_options = state.task_options[task_id]
+        executor = get_step_executor(task_options)
+        metadata_ref, output_ref = executor(
+            state.task_func_body[task_id],
+            state.task_context[task_id],
+            job_id,
+            task_id,
+            baked_inputs,
+            task_options,
+        )
+        # The input workflow is not a reference to an executed workflow.
+        future = asyncio.wrap_future(metadata_ref.future())
+        future.add_done_callback(self._completion_queue.put_nowait)
+
+        state.insert_running_frontier(future, WorkflowRef(task_id, ref=output_ref))
+        state.task_execution_metadata[task_id] = TaskExecutionMetadata(
+            submit_time=time.time()
+        )
+
+    def _post_process_submit_task(
+        self, task_id: StepID, store: "WorkflowStorage"
+    ) -> None:
+        """Update dependencies and reference count etc. after task submission."""
+        state = self._state
+        if task_id in state.continuation_root:
+            if state.task_options[task_id].checkpoint:
+                store.update_continuation_output_link(
+                    state.continuation_root[task_id], task_id
+                )
+        else:
+            # update reference counting if the task is not a continuation
+            for c in state.upstream_dependencies[task_id]:
+                state.reference_set[c].remove(task_id)
+                if not state.reference_set[c]:
+                    del state.reference_set[c]
+                    state.free_outputs.add(c)
+
+    def _garbage_collect(self) -> None:
```

Comment:
Can you add a comment to explain what exactly we are garbage-collecting? I.e. application outputs vs metadata?